core/src/main/java/software/amazon/awssdk/core/async/ByteArrayAsyncRequestBody.java

```diff
@@ -43,20 +45,42 @@ public long contentLength() {
     }
 
     @Override
-    public void subscribe(Subscriber<? super ByteBuffer> subscriber) {
-        subscriber.onSubscribe(
-                new Subscription() {
-                    @Override
-                    public void request(long n) {
-                        if (n > 0) {
-                            subscriber.onNext(ByteBuffer.wrap(bytes));
-                            subscriber.onComplete();
+    public void subscribe(Subscriber<? super ByteBuffer> s) {
+        // As per rule 2.13, we need to throw a `java.lang.NullPointerException` if the `Subscription` is `null`
+        if (s == null) {
+            throw new NullPointerException("Subscription MUST NOT be null.");
+        }
+
+        // As per 2.13, this method must return normally (i.e. not throw).
+        try {
+            s.onSubscribe(
+                    new Subscription() {
+                        boolean done = false;
+
+                        @Override
+                        public void request(long n) {
+                            if (n > 0) {
+                                if (!done) {
+                                    s.onNext(ByteBuffer.wrap(bytes));
+                                    done = true;
+                                    s.onComplete();
+                                }
+                            } else {
+                                s.onError(new IllegalArgumentException("§3.9: non-positive requests are not allowed!"));
+                            }
                         }
-                    }
 
-                    @Override
-                    public void cancel() {
+                        @Override
+                        public void cancel() {
+                        }
                     }
-                });
+            );
+        } catch (Throwable ex) {
+            new IllegalStateException(s + " violated the Reactive Streams rule 2.13 " +
+                    "by throwing an exception from onSubscribe.", ex)
+                    // When onSubscribe fails this way, we don't know what state the
+                    // s is thus calling onError may cause more crashes.
+                    .printStackTrace();
```

Comment:
This seems weird. So we just log the error?

Reply:
I don't think we need to log it but since `onSubscribe` must always return normally unless `s` is null, then we have to swallow it here.

Since 2.13 requires the `Subscriber` implementer to make `onSubscribe` return normally to be in spec, I don't think this will be a big issue.